compiler/semexprs.nim

```diff
@@ -115,15 +117,23 @@ proc checkConvertible(c: PContext, castDest, src: PType): TConvStatus =
     result = convNotLegal
   elif d.kind == tyObject and s.kind == tyObject:
     result = checkConversionBetweenObjects(d, s, pointers)
-  elif (skipTypes(castDest, abstractVarRange).kind in IntegralTypes) and
-      (skipTypes(src, abstractVarRange-{tyTypeDesc}).kind in IntegralTypes):
-    # accept conversion between integral types
-    discard
+  elif (skipTypes(targetTyp, abstractVarRange).kind in IntegralTypes) and
+      (skipTypes(srcTyp, abstractVarRange-{tyTypeDesc}).kind in
+        IntegralTypes + {tyFloat..tyFloat64}):
+    # accept conversion to integral types, unless outside of range
+    if targetTyp.isOrdinalType:
+      if src.kind in nkCharLit..nkUInt64Lit and
+          src.intVal notin firstOrd(targetTyp)..lastOrd(targetTyp):
+        result = convNotInRange
+      elif src.kind in nkFloatLit..nkFloat64Lit and
+          src.floatVal notin
+            firstOrd(targetTyp).toBiggestFloat..lastOrd(targetTyp).toBiggestFloat:
```

Comment:
I've realized that this isn't entirely correct, due to `high(int64).toBiggestFloat` getting rounded up since it's not representable in `float64`. Is there a way to get the largest representable float that is less-or-equal than some integer?

Reply:
Yeah, no idea either, sorry.